### PR TITLE
Test validation success for OpEmitVertex OpEndPrimitive, execution model for primitive instructions

### PR DESCRIPTION
**Pull request by dneto0**

### CODE REVIEW

**Comment by dneto0** [comment]
Followup to #1095 
cc: @jeremy-lunarg

**Comment by dneto0** [comment]
I filed #1098 and #1099 for more validation that can be done for primitive instructions.

**Comment by dneto0** [comment]
Fixed formatting.

File: test/val/val_primitives_test.cpp [hunk 1]
```diff
@@ -28,13 +28,14 @@ using ValidatePrimitives = spvtest::ValidateBase<bool>;
 
 std::string GenerateShaderCode(
     const std::string& body,
+    const bool declare_geometry_streams_capability = true,
     const std::string& capabilities_and_extensions = "",
     const std::string& execution_model = "Geometry") {
```
Comment (atgoo) [inline test/val/val_primitives_test.cpp]: It also should validate and test that it doesn't work when execution model is not Geometry.

Validation can be done calling this function:
    _.current_function().RegisterExecutionModelLimitation(
        SpvExecutionModelGeometry,
        "Primitive instructions require Geometry execution model");
when a primitive instruction is encountered.

File: test/val/val_primitives_test.cpp [hunk 2]
```diff
@@ -28,13 +28,14 @@ using ValidatePrimitives = spvtest::ValidateBase<bool>;
 
 std::string GenerateShaderCode(
     const std::string& body,
+    const bool declare_geometry_streams_capability = true,
```
Comment (atgoo) [inline test/val/val_primitives_test.cpp]: This can be passed through capabilities_and_extensions.

File: test/val/val_primitives_test.cpp (hunk 1 above)
Comment (atgoo) [reply test/val/val_primitives_test.cpp]: (this can be done in the next pull request)

Comment (jeremy-lunarg) [reply test/val/val_primitives_test.cpp]: `_.current_function().RegisterExecutionModelLimitation(
SpvExecutionModelGeometry,
"Primitive instructions require Geometry execution model");`

@atgoo 
I tried to use this naively and failed on `assert(in_function_body())` within `current_function()`.

File: test/val/val_primitives_test.cpp (hunk 2 above)
Comment (dneto0) [reply test/val/val_primitives_test.cpp]: I've reworked it to use that parameter instead.  Good suggestion.

File: test/val/val_primitives_test.cpp (hunk 1 above)
Comment (dneto0) [reply test/val/val_primitives_test.cpp]: I've done this in the current pull request, in the last (new) commit.

**Comment by dneto0** [comment]
All feedback addressed.  Please take another look.

File: test/val/val_primitives_test.cpp [hunk 3]
```diff
@@ -63,6 +63,27 @@ OpFunctionEnd)";
   return ss.str();
 }
 
+// Returns SPIR-V assembly fragment representing a function call,
+// the end of the callee body, and the preamble and body of the called
+// function with the given body, but missing the final return and
+// function-end.  The result is of the form where it can be used in the
+// |body| argument to GenerateShaderCode.
+std::string CallAndCallee(std::string body) {
```
[the remark this answers is not in the document]
Comment (dneto0) [reply test/val/val_primitives_test.cpp]: As I recall the compiler complained when this formal parameter was a const reference.   So I made this pretty dumb by just making non-ref string.  It's dumb and it works.  Clarity is the most important thing in test code, especially at this level of detail.

Thanks for the review.

**Comment by dneto0** [comment]
Rebased, squashed, and merged as  8135dd637

Fixed #1098

**Pull request closed without merging**